### PR TITLE
Feat(toc): Introduce

### PR DESCRIPTION
**Pull request by AugustinMauroy**
## Description

Migration of table of contents from nodejs.dev but using nextra TOC

## Related Issues

#5242 

### Check List

- [X] I have read the [Contributing Guidelines](https://github.com/nodejs/nodejs.org/blob/main/CONTRIBUTING.md) and made commit messages that follow the guideline.
- [X] I have run `npx turbo lint` to ensure the code follows the style guide. And run `npx turbo lint:fix` to fix the style errors if necessary.
- [X] I have run `npx turbo format` to ensure the code follows the style guide.
- [X] I have run `npx turbo test` to check if all tests are passing, and/or `npx turbo test:snapshot` to update snapshots if I created and/or updated React Components.
- [X] I've covered new added functionality with unit tests if necessary.

### CODE REVIEW

File: components/Article/TableOfContents/index.tsx [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+          </li>
+        );
+      })}
+    </ul>
+  );
+};
+const TableOfContents: FC<{ tableOfContents: PropsTableOfContents }> = ({
+  tableOfContents,
+}) => {
+  if (!tableOfContents.length) {
```
Comment (Harkunwar) [inline components/Article/TableOfContents/index.tsx]:
```suggestion
  if (tableOfContents.length === 0) {
```
It's better to have positive checks than negations.

File: components/Article/TableOfContents/index.tsx [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      {tableOfContents.map(item => {
+        return (
+          <li key={item.id}>
+            <Link href={`#${item.id}`}>{item.value}</Link>
+          </li>
+        );
+      })}
+    </ul>
+  );
+};
```
Comment (Harkunwar) [inline components/Article/TableOfContents/index.tsx]:
```suggestion
const traverseTableOfContents = (tableOfContents: PropsTableOfContents) => (
  <ul>
    {tableOfContents.map(item => (
      <li key={item.id}>
        <Link href={`#${item.id}`}>{item.value}</Link>
      </li>
    ))}
  </ul>
);

```
You don't need `return` statements. 


Also, can you convert this to a Component instead of a Function. It's considered a bad pattern.

Comment (AugustinMauroy) [reply components/Article/TableOfContents/index.tsx]: Thanks for giving this advidce but on original component we have nested list.
I have to update the component

Comment (AugustinMauroy) [reply components/Article/TableOfContents/index.tsx]: Now take a look, `return` is needed

Comment (ovflowd) [reply components/Article/TableOfContents/index.tsx]: Why is return needed?

Comment (AugustinMauroy) [reply components/Article/TableOfContents/index.tsx]: Look up the code now, you will have the answer 😁

File: components/Article/TableOfContents/index.tsx [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+      {MappedTableOfContents.map(item => (
+        <li key={item.id}>
+          <Link href={`#${item.id}`}>{item.value}</Link>
+          {item.children && traverseTableOfContents(item.children)}
+        </li>
+      ))}
+    </ul>
+  );
+};
+const TableOfContents: FC<{ tableOfContents: PropsTableOfContents }> = ({
```
Comment (shanpriyan) [inline components/Article/TableOfContents/index.tsx]:
```suggestion
};

const TableOfContents: FC<{ tableOfContents: PropsTableOfContents }> = ({
```

**Review by shanpriyan — CHANGES_REQUESTED** [review]

File: components/Article/TableOfContents/index.tsx (hunk 2 above)
Comment (shanpriyan) [reply components/Article/TableOfContents/index.tsx]: @AugustinMauroy As Harkunwar mentioned this function can be changed to React component

Comment (shanpriyan) [reply components/Article/TableOfContents/index.tsx]: I don't see a problem of using `return` here but if you really want to remove it you can change the following line
```diff
- {MappedTableOfContents.map(item => (
+ {parseTableOfContents(tableOfContents).map(item => (
```

Comment (AugustinMauroy) [reply components/Article/TableOfContents/index.tsx]: Good idea !

**Comment by ovflowd** [comment]
@AugustinMauroy the builds are failing, just as a fyi

**Comment by AugustinMauroy** [comment]
I know it's from type
i have to take time for this

**Comment by mikeesto** [comment]
Hi @AugustinMauroy . I think if you update the storybook snapshot now, the build will be successful

**Comment by AugustinMauroy** [comment]
@mikeesto that an point but they are an other big part. It's typing of parsing function
BTW thanks

**Comment by ovflowd** [comment]
@AugustinMauroy I didn't get your reply to @mikeesto could you rephrase?

**Comment by mikeesto** [comment]
I think I understand now. There are two issues: (1) storybook snapshot needs updating, and (2) the types in `TableOfContents/index.tsx` could be improved. This is how I would type it @AugustinMauroy . Feel free to take it as is or modify.

```tsx

type TableOfContentsNode = {
  depth: number;
  value: string;
  id: string;
  children?: TableOfContentsNode[];
  parent?: TableOfContentsNode | null;
};

type TableOfContentsProps = {
  tableOfContents: TableOfContentsNode[];
};

function parseTableOfContents(
  tableOfContents: TableOfContentsNode[]
): TableOfContentsNode[] {
  const parsedTableOfContents: TableOfContentsNode[] = [];

  function addChildNode(
    parent: TableOfContentsNode,
    node: TableOfContentsNode
  ) {
    if (!parent.children) {
      parent.children = [];
    }
    parent.children.push(node);
  }

  tableOfContents.forEach(node => {
    const { depth, value, id } = node;
    const parsedNode: TableOfContentsNode = { depth, value, id };

    let parentNode = parsedTableOfContents[parsedTableOfContents.length - 1];
    while (parentNode && parentNode.depth >= depth) {
      parentNode = parentNode.parent!;
    }

    if (parentNode) {
      addChildNode(parentNode, parsedNode);
    } else {
      parsedTableOfContents.push(parsedNode);
    }

    parsedNode.parent = parentNode || null;
  });

  return parsedTableOfContents;
}

const traverseTableOfContents: FC<TableOfContentsNode[]> = tableOfContents => (
  <ul>
    {parseTableOfContents(tableOfContents).map(item => (
      <li key={item.id}>
        <Link href={`#${item.id}`}>{item.value}</Link>
        {item.children && traverseTableOfContents(item.children)}
      </li>
    ))}
  </ul>
);

const TableOfContents: FC<TableOfContentsProps> = ({ tableOfContents }) => {
  if (tableOfContents.length === 0) {
    return null;
  }
  return (
    <details className={styles.tableOfContents}>
      <summary>
        <strong>
          <FormattedMessage id="components.article.tableOfContents" />
        </strong>
      </summary>
      {traverseTableOfContents(tableOfContents)}
    </details>
  );
};
```

**Comment by AugustinMauroy** [comment]
@mikeesto Yep i will test you code. Thanks 😄!

**Comment by ovflowd** [comment]
**FYI:** We deleted the `major/website-redesign` branch to simplify our release and rollout process. Please feel free to recreate this PR or reopen it with the `main` branch as a target.

Thank you, and we apologise for the inconvenience!
